src/material/datepicker/calendar.ts

```diff
@@ -255,19 +256,43 @@ export class MatCalendar<D> implements AfterContentInit, AfterViewChecked, OnDes
     return this._minDate;
   }
   set minDate(value: D | null) {
-    this._minDate = this._dateAdapter.getValidDateOrNull(this._dateAdapter.deserialize(value));
+    this._oldMinDate = this._minDate;
+
+    const validated = this._dateAdapter.getValidDateOrNull(this._dateAdapter.deserialize(value));
+    if (validated === null) {
+      this._minDate = null;
+    } else {
+      this._minDate = this._dateAdapter.createDate(
+        this._dateAdapter.getYear(validated),
+        this._dateAdapter.getMonth(validated),
+        this._dateAdapter.getDate(validated),
+      );
+    }
   }
   private _minDate: D | null;
+  private _oldMinDate: D | null;
 
   /** The maximum selectable date. */
   @Input()
   get maxDate(): D | null {
     return this._maxDate;
   }
   set maxDate(value: D | null) {
-    this._maxDate = this._dateAdapter.getValidDateOrNull(this._dateAdapter.deserialize(value));
+    this._oldMaxDate = this._maxDate;
+
+    const validated = this._dateAdapter.getValidDateOrNull(this._dateAdapter.deserialize(value));
```

Comment:
Is it possible to move this logic into a separate function? We seem to be repeating this in minDate's setter